### DISCUSSION TITLE
飞行时双击向下时不能降落

### DISCUSSION THREAD

**Discussion by WhiteElephant-abc** [TouchController/TouchController#116] [输入控制逻辑]
在国际基岩版1.21.51.02中，双击向上和向下都能退出飞行

**Comment by fengzhou0w0** [comment]
实测旧版（pe版）没有这个特性

**Comment by WhiteElephant-abc** [comment]
TC新控件贴图都有，你总不能让它对标旧版吧

**Comment by fengzhou0w0** [comment]
> TC新控件贴图都有，你总不能让它对标旧版吧

忽然感觉，新版控件的“上升下降”与“跳跃潜行”按键之间强绑定，和老版控件四个按键分开控制可以说逻辑完全不同了。所以我在想，与其让TC模组一套触控兼顾所有基岩版的特性，不如分成新旧两套供玩家选择。而不是只是贴图不同。因为仔细研究感觉新旧控件之间还是有逻辑冲突的。

**Comment by WhiteElephant-abc** [comment]
同意，例如 #104 中就是这个问题

**Comment by WhiteElephant-abc** [comment]
> > TC新控件贴图都有，你总不能让它对标旧版吧
> 
> 忽然感觉，新版控件的“上升下降”与“跳跃潜行”按键之间强绑定，和老版控件四个按键分开控制可以说逻辑完全不同了。所以我在想，与其让TC模组一套触控兼顾所有基岩版的特性，不如分成新旧两套供玩家选择。而不是只是贴图不同。因为仔细研究感觉新旧控件之间还是有逻辑冲突的。

我觉得在设置里添加一个开关，可以选择新旧逻辑，这样就挺好

**Comment by WhiteElephant-abc** [comment]
> > > TC新控件贴图都有，你总不能让它对标旧版吧
> > 
> > 
> > 忽然感觉，新版控件的“上升下降”与“跳跃潜行”按键之间强绑定，和老版控件四个按键分开控制可以说逻辑完全不同了。所以我在想，与其让TC模组一套触控兼顾所有基岩版的特性，不如分成新旧两套供玩家选择。而不是只是贴图不同。因为仔细研究感觉新旧控件之间还是有逻辑冲突的。
> 
> 我觉得在设置里添加一个开关，可以选择新旧逻辑，这样就挺好

但是新逻辑比旧的少两个控件啊...

**Comment by fengzhou0w0** [comment]
确实，仔细想想也不用强行分成两套。只需要给十字键的中心键再增加两套可选的不同功能，然后再增加两个独立按钮就行。

方案：

十字键中心键增加后的功能：
1.潜行
2.跳跃
3.潜行/下降（双击切换）
4.跳跃/上升（双击切换）

增加独立按钮：
1.潜行/下降（双击切换）
2.跳跃/上升（双击切换）

**Comment by fengzhou0w0** [comment]
如果增加这两个独立按键，默认贴图可以用新版贴图，用来在选按钮的时候与旧按钮进行区分。

**Comment by fengzhou0w0** [comment]
梳理一下新旧按钮的逻辑
旧版本按钮逻辑：
1.单击跳跃键：跳跃
2.双击潜行键：潜行
3.双击跳跃键：飞行，显示起飞与降落键

新版本按钮逻辑：
1.单击跳跃键：跳跃
2.双击跳跃键：切换为起飞键
3.双击起飞键：切换为跳跃键
4.单击潜行键：潜行
5.双击潜行键：切换为降落键
6.双击降落键：切换为潜行键

现模组版本的键位是旧版逻辑，你用着感觉很难受，但如果按你的请求全部改为新版逻辑，那我会很难受。所以我个人想到的最好的方案就是上面的那条发言。不过我个人还是以作者大大的想法为准，我只是提个参考。

**Comment by fifth-light** [comment]
> TC新控件贴图都有，你总不能让它对标旧版吧

默认贴图是老版贴图，我是打算先做出老版的控制的。我认为新老版本各种差异过大，需要集结为一个专门收集这些小行为的集中反馈区，而不是单独提 issue。

> 只需要给十字键的中心键再增加两套可选的不同功能，然后再增加两个独立按钮就行。

> 现模组版本的键位是旧版逻辑，你用着感觉很难受，但如果按你的请求全部改为新版逻辑，那我会很难受。

把不同按钮逻辑做成可选择的编辑项这一点上我比较赞同。未来的 TouchController 模组会有一个图层系统，可以根据不同的情况显示不同的图层。目前的版本根据不同的情况显示不同的贴图或者显示不同的按键只是一个临时策略。有了图层系统之后就可以实现这些需求了，但是图层系统我打算等新的界面库做好后再添加相关支持。

**Comment by fifth-light** [comment]
> 需要集结为一个专门收集这些小行为的集中反馈区，而不是单独提 issue。

关于这一点 @WhiteElephant-abc @fengzhou0w0 你们觉得怎么样？

**Comment by fengzhou0w0** [comment]
![Screenshot_2024-12-22-23-11-14-094_com.mojang.minecraftpe.jpg](https://github.com/user-attachments/assets/61ed6db2-aef1-447c-9055-48fd9f133ab6)

不对，刚才测试了一下摇杆控件，我提出的有些机制说法是错误的。不过话说回来，想实现你的想法还是得靠增加新按键。（比如右边这三个按键，表面上只增加了疾跑键，实际上三个按键的机制似乎都跟旧版不一样）

**Reply by WhiteElephant-abc** [reply]
想问一下那个箭头是什么按键

**Reply by fengzhou0w0** [reply]
进入疾跑状态的按键。你玩基岩版应该是知道的吧

**Reply by WhiteElephant-abc** [reply]
不常玩其实

**Reply by WhiteElephant-abc** [reply]
> 进入疾跑状态的按键。你玩基岩版应该是知道的吧

好像还有一个两个方框叠起来，再有一个加号，这个是什么按键？

**Reply by fengzhou0w0** [reply]
> 好像还有一个两个方框叠起来，再有一个加号，这个是什么按键？

没有印象，你要不给个图

**Reply by fifth-light** [reply]
> 想问一下那个箭头是什么按键

因为你开了“使用摇杆疾跑”，打开这个选项时会隐藏疾跑键。

**Comment by fengzhou0w0** [comment]
> > 需要集结为一个专门收集这些小行为的集中反馈区，而不是单独提 issue。
> 
> 关于这一点 @WhiteElephant-abc @fengzhou0w0 你们觉得怎么样？

赞成，今日的issue大多都是这类问题，开一个集中反馈区对梳理问题更有帮助。还有关于issue的讨论部分不知道能不能移到discussions板块，issue仅保留请求部分。

**Comment by fifth-light** [comment]
> 关于issue的讨论部分不知道能不能移到discussions板块，issue仅保留请求部分。

我觉得要不先在 Discussion 讨论，然后确定具体的实现方法，最后再创建相关的 issue 跟进进度，这样会更好一些。

**Comment by fengzhou0w0** [comment]
> 我觉得要不先在 Discussion 讨论，然后确定具体的实现方法，最后再创建相关的 issue 跟进进度，这样会更好一些。

也好，未来我个人有想法就去discussion提出吧，如果你认为有实现的价值，你再搬到issue上。

**Comment by fifth-light** [comment]
@WhiteElephant-abc 你觉得怎么样，如果行的话我就把与基岩版实现不一致的 issue 迁移到 dicussion 了。

**Comment by fifth-light** [comment]
> 需要集结为一个专门收集这些小行为的集中反馈区

至于这一点，我认为可以做一个关于基岩版各种控制行为的 wiki，专门收集这些操作行为和实现细节

**Comment by WhiteElephant-abc** [comment]
> @WhiteElephant-abc 你觉得怎么样，如果行的话我就把与基岩版实现不一致的 issue 迁移到 dicussion 了。

我觉得可以

**Comment by fifth-light** [comment]
事实上现在向下飞纯纯就是对应shift。因此双击shift肯定不会触发取消飞行。真要实现我得重做这些按键的逻辑了。
